### ISSUE TITLE
Explore the possibility of load balancing RDS read replicas

### ISSUE THREAD

**Issue by rjayasekera** [fecgov/openFEC#3945]
We would like to have application servers send queries to the least busy read replica.  Currently we have a environment variable (SQLA_FOLLOWERS) specify connections to both read read replicas.  

It looks like application server randomly pick which databases instance to send requests.  

We would like applications servers to figure out which database is the least busy first then send the queries to that instance.

In short placing a load balance in-front of AWS RDS read replicas

- Scaling Your Amazon RDS Instance Vertically and Horizontally
https://aws.amazon.com/blogs/database/scaling-your-amazon-rds-instance-vertically-and-horizontally/

- How can I distribute read requests across multiple Amazon RDS read replicas?
https://aws.amazon.com/premiumsupport/knowledge-center/requests-rds-read-replicas/

**Comment by fecjjeng** [comment]
We had analyzed the recent slow queries, the major problems all involved schedule A related endpoints.

**Comment by fecjjeng** [comment]
We had hold a workshop with experts from AWS on 11/15/2019 to look into the performance issues we had.  We had provided some queries that we discussed during the workshop.  Even SQLAlchemy generated queries are usually not most optimized for performance, these sets of queries generated for the schedule_A endpoint are pretty straightforwards.  The bottleneck appears to be on the size of schedule_a tables, both on the number of rows, and the number of columns.
Following is the number of rows for partitioned schedule_a tables after 1990:
fec_fitem_sched_a_1991_1992    886900
fec_fitem_sched_a_1993_1994    839804
fec_fitem_sched_a_1995_1996    1234943
fec_fitem_sched_a_1997_1998    1016239
fec_fitem_sched_a_1999_2000    1698031
fec_fitem_sched_a_2001_2002    1443090
fec_fitem_sched_a_2003_2004    6841837
fec_fitem_sched_a_2005_2006    6748247
fec_fitem_sched_a_2007_2008    12702280
fec_fitem_sched_a_2009_2010    9579001
fec_fitem_sched_a_2011_2012    20975518
fec_fitem_sched_a_2013_2014    22811198
fec_fitem_sched_a_2015_2016    55582590
fec_fitem_sched_a_2017_2018    88743635
fec_fitem_sched_a_2019_2020    23591865

**Comment by fecjjeng** [comment]
The AWS team had sent a list of recommendation afterwards that they think may help:
After having a chance to dive deeper into the explain plan, review the DDL, and consult with some colleagues, I have a few suggestion you might want to consider.


CONTEXT
1) There is CASTING going on which is due to the NUMERIC data types that probably should be INTEGER data types.
2) Data types that should be NUMERIC data types need PRECISION and SCALE.
2) The section below shows that 440,700 rows are being retrieved via the predicate, which is fine, except that the blocks are rescanned to validate "is_indivdual" and "two_year_transaction_period" is in each block so the block scans are almost doubled.
--> Bitmap Heap Scan on fec_fitem_sched_a_2017_2018  (cost=4130.87..1595965.53 rows=405488 width=78)
Recheck Cond: (contributor_occupation_text @@ to_tsquery('lawyer:*'::text))'
           Filter: (is_individual AND (two_year_transaction_period = '2018'::numeric))'
                -->  Bitmap Index Scan on idx_sched_a_2017_2018_contrib_occ_text_dt_sub_id  (cost=0.00..4029.50 rows=440700 width=0)'
                             Index Cond: (contributor_occupation_text @@ to_tsquery('lawyer:*'::text))'


DATA TYPES
contbr_zip character varying,
--> can this be an integer?
fec_election_yr character varying,
--> can this be an integer?
contb_aggregate_ytd numeric,
--> numeric(11,2) Does the precision and scale need to be more than 000000000.00 since it's a monetary value?
contb_receipt_amt numeric,
--> numeric(11,2) Does the precision and scale need to be more than 000000000.00 since it's a monetary value?
conduit_cmte_zip character varying,
--> can this be an integer?
increased_limit character varying,
--> numeric(11,2)? Does the precision and scale need to be more than 000000000.00 since it's a monetary value?
line_num character varying,
--> can this be an integer?
image_num character varying,
--> can this be an integer?
file_num numeric,
--> integer
link_id numeric,
--> integer
orig_sub_id numeric,
--> integer
sub_id numeric,
--> integer
rpt_yr numeric,
--> integer
two_year_transaction_period  numeric, --> integer

ALTER TABLE disclosure.fec_fitem_sched_a_2019_2020
ALTER COLUMN contbr_zip                                    TYPE INT,
ALTER COLUMN fec_election_yr                           TYPE INT,
ALTER COLUMN contb_aggregate_ytd                TYPE NUMERIC(11,2),
ALTER COLUMN contb_receipt_amt                    TYPE NUMERIC(11,2),
ALTER COLUMN conduit_cmte_zip                      TYPE INT,
ALTER COLUMN increased_limit                          TYPE NUMERIC(11,2),
ALTER COLUMN line_num                                     TYPE INT,
ALTER COLUMN image_num                                 TYPE INT,
ALTER COLUMN file_num                                      TYPE INT,
ALTER COLUMN link_id                                          TYPE INT,
ALTER COLUMN orig_sub_id                                 TYPE INT,
ALTER COLUMN sub_id                                          TYPE INT,
ALTER COLUMN rpt_yr                                           TYPE INT,
ALTER COLUMN two_year_transaction_period TYPE INT;


INDEXES
CREATE INDEX idx_sched_a_2019_2020_two_year_period_<name>
  ON disclosure.fec_fitem_sched_a_2019_2020 USING GIN
  (contributor_occupation_text, two_year_transaction_period, is_individual);

CREATE INDEX idx_sched_a_2019_2020_two_year_period_<name>
  ON disclosure.fec_fitem_sched_a_2019_2020 USING BTREE
  (cmte_id, two_year_transaction_period);

CREATE INDEX idx_sched_a_2019_2020_two_year_period_<name>
  ON ofec_committee_history_mv USING BTREE
  (committee_id, cycle);

CREATE INDEX idx_sched_a_2019_2020_two_year_period_<name>
  ON disclosure.fec_fitem_sched_a_2019_2020 USING BTREE   
  (clean_contbr_id, two_year_transaction_period);

CREATE INDEX idx_sched_a_2019_2020_two_year_period_<name>
  ON ofec_committee_history_mv USING BTREE
  (committee_id, cycle);


REDSHIFT
Because the amount of data is going to continue to grow and the access patterns are dynamic, Redshift is going to be the right tool for the job for long-term sustainability because the data can be shaped specifically to support the access pattern (fit-for-purpose).

**Comment by fecjjeng** [comment]
The casting would cost for performance.  However, the data_type recommendation they suggested that could be integer are either varchar in nature (such as zip code, line num, image num) or is actually a string of number in nature and too big to fit into integer (such as sub_id, link_id, orig_sub_id).  So they are not relevant in our case.

**Comment by fecjjeng** [comment]
For the indexes they suggested, we did not create index on is_individual is because the cardinality is very low for it, true or false.  With the majority of the rows with value true.  The earlier tests we had conducted indicated even with the existence of the index, the optimizer does not use it.  The two_year_transaction_period is the partition key, i.e., the entire partition table has the same two_year_transaction_period.

**Comment by hcaofec** [comment]
As for zip codes, there are foreign zip codes in our tables which are characters.

**Comment by fecjjeng** [comment]
> As for zip codes, there are foreign zip codes in our tables which are characters.

**Comment by fecjjeng** [comment]
This is the last suggestion from AWS expert:
Do you this specific index?

CREATE INDEX idx_sched_a_2019_2020_two_year_period_<name>
 ON disclosure.fec_fitem_sched_a_2019_2020 USING GIN
 (contributor_occupation_text, two_year_transaction_period, 
is_individual);

**Comment by fecjjeng** [comment]
We had tested the above suggestion.  Following is the test results:

We currently don’t have composite index of this combination.  

As shown in the ddl included in the previous email, we do have Gin index on contributor_occupation_text and btree index on two_year_transaction_period, combined with contb_receipt_amt, sub_id/ and contb_receipt_dt, sub_id, for sorting/pagination purpose.

But to test your suggestion, we created the following index on one partition table fec_fitem_sched_a_2019_2020.
CREATE INDEX idx_sched_a_2019_2020_contrib_occ_text_cycle
 ON disclosure.fec_fitem_sched_a_2019_2020 USING GIN
 (contributor_occupation_text, two_year_transaction_period);

is_individual is of boolean data_type and has no default operator class for access method “gin” so can not be included in the new GIN index

The query we used to test this new index is as followed:
select * 
FROM disclosure.fec_fitem_sched_a 
WHERE disclosure.fec_fitem_sched_a.contributor_occupation_text @@ to_tsquery('lawyer:*') 
AND disclosure.fec_fitem_sched_a.is_individual = true 
AND disclosure.fec_fitem_sched_a.two_year_transaction_period = 2020
--ORDER BY disclosure.fec_fitem_sched_a.contb_receipt_dt DESC, disclosure.fec_fitem_sched_a.sub_id DESC LIMIT 30;

## ###########
## with our current indexes 
## ###########
-- without sort
-- 111647 rows, 2:14 minutes

-- with sort/limit 30
-- 30 rows.  353 msec

## ###########
## with added new index (contributor_occupation_text, two_year_transaction_period)
## ###########
-- without sort
— the new index had been chosen by the optimizer.  However, with similar return time
-- 111647 rows, 2:12 minutes

-- with sortt/limit 30
-- 30 rows.  367 msec

----------------
An additional gin index on contributor_occupation_text only had also been created to check the effect
----------------
CREATE INDEX idx_sched_a_2019_2020_contrib_occ_text
 ON disclosure.fec_fitem_sched_a_2019_2020 USING GIN
 (contributor_occupation_text);  
----------------

## ###########
## with added new index (contributor_occupation_text, two_year_transaction_period)
## with added new index (contributor_occupation_text)
## ###########
-- without sort
— the new index (contributor_occupation_text) had been chosen by the optimizer.  However, with similar return time
-- 111647 rows, 2:12 minutes

-- with sortt/limit 30
-- 30 rows.  357 msec

**Comment by fecjjeng** [comment]
AWS/Rohan had set up a briefing session on 12/6/2019.
An AWS Solution Architect with significant RedShift experience will have a briefing on AWS Redshift.
